Extract sub-object in interactive mode

crackleware opened the pull request:
When sub-object is selected in interactive mode, by pressing `x` or `Ctrl-e` sub-object is printed to stdout in this form:
```
$ <t2.json fx
{
  "a": 11,
  "b": {
    "c": 33
  }
}

$ <t2.json fx # .b is interactively selected, key `x` is pressed
{
  "path": ".b",
  "value": {
    "c": 33
  }
}
```

I had to do some changes around handling piped/TTY input and output and command line arguments. 
It should support all existing modes of use.

This should also close #132

Best regards!

_UPDATE:_ I removed anything about completely unneeded interactive mode forcing flags. PR is now only about interactive mode action for extraction of sub-object to support semi-interactive workflows discussed in the PR.

crackleware commented:
I'll look into failing tests tomorrow.

antonmedv commented:
I don't get it. For interactive/noninteractive mode do this:
```
cat ... | fx .filed | fx
```
```
cat ... | fx .
```

For the selected path, yes. This is something to consider.

crackleware commented:
> I don't get it. For interactive/noninteractive mode do this:
> 
> ```
> cat ... | fx .filed | fx
> ```
> 
> ```
> cat ... | fx .
> ```

Ah I see your point. Got it.

> For the selected path, yes. This is something to consider.

Yeah, there are some issues. For example, I only use changes to `fx.js`, and `index.js`, where stdin/stdout and arguments handling is done, I keep unchanged (from master branch).

```
$ <t2.json fx .
{
  "a": 11,
  "b": {
    "c": 22,
    "d": {
      "e": "final destination"
    }
  }
}
$ <t2.json fx .b | fx | fx .value.e
undefined
```

In second `fx` in the pipeline I want to interactively select `.d` and then the next command in pipeline `fx .value.e` should extract `final destination`.

How we should do this? Any ideas?

crackleware commented:
Of course, I don't see how it will be possible to compose multiple interactive selections in one pipeline. Those will have to be broken, for example, like this:
```
$ path1="$(<t2.json fx .b | fx | fx .path)"
$ ...do-something-with-$path1
$ <some.json | fx $path1 ... | fx | ...
```
All of this can be in one line, separated by `;`.

crackleware commented:
And something like this will probably be okay:
```
$ cat t2.json
{
    "a": 11,
    "b": {
        "c": [
            {"f": 1},
            {"f": 2},
            {"f": 3}
        ],
        "d": {
            "e": "final destination"
        }
    }
}
$ <t2.json fx "$(<t2.json fx | fx .path | sed 's/\[.*\]/[]/g')"
```
We first select some sub-object from `t2.json`, get path to it, then transform path, then use transformed path to extract sub-objects from `t2.json`.

crackleware commented:
Another alternative to above, with more interactivity regarding path transformation:
```
$ <t2.json fx $(<t2.json fx | fx .path | vipe)
```
Using `vipe` tool (from https://joeyh.name/code/moreutils) to interactively edit JSON sub-object path in your favorite editor.

But I don't want to do editing everytime if I don't need to so I'm thinking about set of commands in the shell for semi-interactive computational pipeline/workflow data management. Maybe if we prefix commands with functional purity decorators we get transparent caching of previous outputs, and interactively specify new inputs only for parts of DAG that is being recomputed. But this is probably OT. Thanks for `fx` and inspiration 😄

crackleware commented:
One more alternative workflow. Looks flexible. TAB expands back-quoted commands. So,

![fx-subobj-path-extraction-workflow-1](https://user-images.githubusercontent.com/662771/92984805-57eeaa80-f4ad-11ea-87f8-9806d4816dd2.gif)

antonmedv commented:
Great ideas! 💡 Mixing interactive is always difficult. Also I’m thinking to rewrite part of fx what is responsible for interactive part. As currently it uses blessed nom package and it has lots of bugs in it. I want to rewrite it from scratch.

crackleware commented:
Cool! 👍  I don't have experience with blessed library in nodejs so I trust you 💯 % 

I simplified and cleaned up this PR. Removed anything about unneeded interaction mode forcing flags. Thanks for clarifying that. Tests pass.

But you'll be the judge 😃  when/if to merge it given your development plans regarding awesome interactive mode. At least we are exploring possibilities in this PR.

I was casually browsing through recently pushed forks and I like this idea too: 
**"Add "c" keybind to copy current path to clipboard"** https://github.com/brewingcode/fx/commit/53b2f0f3bc5b9c5371fed07485db3c70955b5b64  

Maybe it's better to make extensible mechanism for user-defined actions which will pipe currently selected sub-object to stdin of sub-process (configured in `.fxrc` and/or environment variables and/or command line args). There could be multiple such user-defined actions. If you think it's good idea I may create new PR for that.

crackleware commented:
Streaming doesn't work with changes in this PR currently. Looking into it.

crackleware commented:
> Streaming doesn't work with changes in this PR currently. Looking into it.

Fixed.

antonmedv commented:
I rewrote the entire fx in golang. So I belove this PR is no longer actual.
